torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,715 @@
+#include "../Cuda.hpp"
+#include "../../../csrc/utils/auto_gpu.h"
+#include "DataChannelNccl.hpp"
+#include "DataChannelUtils.hpp"
+
+#include <cuda.h>
+#include <THC/THC.h>
+
+#include <unistd.h>
+
+#include <cstdint>
+#include <stdexcept>
+#include <unordered_set>
+#include <sstream>
+
+namespace thd {
+
+namespace {
+
+
+std::unordered_map<THDReduceOp, ncclRedOp_t> ncclOp = {
+  {THDReduceOp::THDReduceMIN, ncclMin},
+  {THDReduceOp::THDReduceMAX, ncclMax},
+  {THDReduceOp::THDReduceSUM, ncclSum},
+  {THDReduceOp::THDReducePRODUCT, ncclProd},
+};
+
+
+std::unordered_map<at::ScalarType, ncclDataType_t> ncclDatatype = {
+  {at::kChar, ncclInt8},
+  {at::kByte, ncclUint8},
+  {at::kFloat, ncclFloat},
+  {at::kDouble, ncclDouble},
+  {at::kInt, ncclInt32},
+  {at::kLong, ncclInt64},
+};
+
+
+// Helper function that gets the data type and issues error if not supported
+static ncclDataType_t _getNcclDataType(at::ScalarType type) {
+  if (ncclDatatype.find(type) == ncclDatatype.end()) {
+    throw std::runtime_error("Unsupported data type for NCCL");
+  }
+  return ncclDatatype[type];
```

Comment:
nit: `find` gives you an iterator that will give you the answer immediately.
I think it's nicer to rewrite this as that:
```cpp
try {
  return ncclDatatype.at(type);
} catch (std::out_of_range& e) {
  throw std::runtime_error("Unsupported data type for NCCL backend");
}
```